src/scripts/main.js

```diff
@@ -1,3 +1,332 @@
 'use strict';
 
-// write your code here
+const gameField = document.querySelector('.game-field');
+const button = document.querySelector('button');
+const allCells = gameField.querySelectorAll('td');
+const allRows = gameField.querySelectorAll('tr');
+const gameScore = document.querySelector('.game-score');
+const messageBox = document.querySelector('.message-container');
+let movesCheckerFlag;
+
+for (const cell of allCells) {
+  cell.setAttribute('empty', true);
+}
+
+function generateMatrix() {
+  const matrix = [[], [], [], []];
+
+  for (let i = 0; i < 4; i++) {
+    for (let j = 0; j < 4; j++) {
+      matrix[i].push(+allRows[i].children[j].innerText);
+    }
+  }
+
+  return matrix;
+}
+
+function transferMatrixOnField(matrix) {
+  const cells = gameField.querySelectorAll('td');
+  const array = matrix.flat();
+
+  for (let i = 0; i < 16; i++) {
+    cells[i].innerText = !array[i] ? '' : array[i];
+    cells[i].className = '';
+    cells[i].classList.add('field-cell', `field-cell--${array[i]}`);
+
+    if (!cells[i].innerText) {
+      cells[i].setAttribute('empty', true);
+    } else {
+      cells[i].setAttribute('empty', false);
+    }
+  }
+}
+
+function checkWin() {
+  const fieldCells = gameField.querySelectorAll('td');
+
+  for (const cell of fieldCells) {
+    if (cell.innerText === '2048') {
+      messageBox.children[1].classList.remove('hidden');
+      break;
+    }
+  }
+}
+
+function checkMoves() {
+  const matrix = generateMatrix();
+
+  for (let i = 0; i < 4; i++) {
+    for (let j = 0; j < 4; j++) {
+      if (matrix[i][j] === 0) {
+        movesCheckerFlag = true; break;
+      } else if (i < 3 && matrix[i][j] === matrix[i + 1][j]) {
+        movesCheckerFlag = true; break;
+      } else if (j < 3 && matrix[i][j] === matrix[i][j + 1]) {
+        movesCheckerFlag = true; break;
+      } else {
+        movesCheckerFlag = false;
+      }
+    }
+
+    if (movesCheckerFlag) {
+      break;
+    }
+  }
+
+  if (!movesCheckerFlag) {
+    gameOver();
+  }
+}
+
+function gameOver() {
+  messageBox.children[0].classList.remove('hidden');
+}
+
+function generateTwoFour() {
+  const array = [2, 2, 2, 2, 2, 2, 2, 2, 2, 4];
+  const index = Math.round(Math.random() * 9);
+
+  return array[index];
+}
+
+function findRandomEpmtyCell(epmtyCellsQuantity) {
+  return Math.floor(Math.random() * epmtyCellsQuantity);
+}
+
+function generateNewNumberOnField() {
+  const emptyCells = gameField.querySelectorAll('[empty="true"]');
```

Comment:
```suggestion
  const emptyCells = gameField.querySelectorAll('[empty="true"]');
```
знову ж таки, погана ідея довіряти DOM, в тебе є матриця з ігровим полем як я зрозумів, якщо ні, то тоді має бути, зберігай дані в ній а не в DOM, DOM просто для рендеру і показів результатів, вся ігрова логіка і механіка зберігається тут в JS